### PR TITLE
Fix the issue that the VMSS upgrading doesn't work if the topology do…

### PR DESCRIPTION
**Pull request by chengliangli0918**
…esn't have master nodes

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/operations/kubernetesupgrade/upgrader.go [hunk 1]
```diff
@@ -383,6 +385,43 @@ func (ku *Upgrader) upgradeAgentPools(ctx context.Context) error {
 }
 
 func (ku *Upgrader) upgradeAgentScaleSets(ctx context.Context) error {
+	if len(ku.ClusterTopology.AgentPoolScaleSetsToUpgrade) != 0 {
```
Comment (jackfrancis) [inline pkg/operations/kubernetesupgrade/upgrader.go]: This should probably be `> 0`, right?

Comment (chengliangli0918) [reply pkg/operations/kubernetesupgrade/upgrader.go]: thanks for the comments, I will change it to > 0 in next iteration.

File: pkg/operations/kubernetesupgrade/upgrader.go [hunk 2]
```diff
@@ -383,6 +385,43 @@ func (ku *Upgrader) upgradeAgentPools(ctx context.Context) error {
...
+		}
+
+		if err := transformer.NormalizeForVMSSScaling(ku.logger, templateMap); err != nil {
+			ku.logger.Errorf("unable to update template, error: %v.", err)
+			return err
+		}
+
+		random := rand.New(rand.NewSource(time.Now().UnixNano()))
+		deploymentSuffix := random.Int31()
+		deploymentName := fmt.Sprintf("agentscaleset-%s-%d", time.Now().Format("06-01-02T15.04.05"), deploymentSuffix)
```
Comment (CecileRobertMichon) [inline pkg/operations/kubernetesupgrade/upgrader.go]: `time.Now().Format("06-01-02T15.04.05")` why is the timestamp hardcoded here?

File: pkg/operations/kubernetesupgrade/upgrader.go [hunk 3]
```diff
@@ -383,6 +385,43 @@ func (ku *Upgrader) upgradeAgentPools(ctx context.Context) error {
 }
 
 func (ku *Upgrader) upgradeAgentScaleSets(ctx context.Context) error {
+	if len(ku.ClusterTopology.AgentPoolScaleSetsToUpgrade) > 0 {
+		// need to apply the ARM template with target kubernets version to the VMSS first in order that the new VMSS instances
```
Comment (CecileRobertMichon) [inline pkg/operations/kubernetesupgrade/upgrader.go]: kubernets --> Kubernetes

Comment (chengliangli0918) [reply pkg/operations/kubernetesupgrade/upgrader.go]: Nice catch, thanks. Fix it in commit 3.

File: pkg/operations/kubernetesupgrade/upgrader.go (hunk 2 above)
Comment (chengliangli0918) [reply pkg/operations/kubernetesupgrade/upgrader.go]: This is not hardcoded timestamp, it defines a format to change the time.Now() to string, e.g. 18-08-12T22.33.17. 
This new line just follows the existing pattern used on Line 59 in pkg\operations\kubernetesupgrade\upgradeagentnode.go

deploymentName := fmt.Sprintf("master-%s-%d", time.Now().Format("06-01-02T15.04.05"), deploymentSuffix)

File: pkg/operations/kubernetesupgrade/upgrader.go [hunk 4]
```diff
@@ -383,6 +385,43 @@ func (ku *Upgrader) upgradeAgentPools(ctx context.Context) error {
 }
 
 func (ku *Upgrader) upgradeAgentScaleSets(ctx context.Context) error {
+	if len(ku.ClusterTopology.AgentPoolScaleSetsToUpgrade) > 0 {
+		// need to apply the ARM template with target Kubernetes version to the VMSS first in order that the new VMSS instances
+		// created can get the expected Kubernetes version. Otherwise the new instances created still have old Kubernetes version
+		// if the topology dosn't have master nodes(so there is no ARM deployments in previous upgradeMasterNodes step)
```
Comment (CecileRobertMichon) [inline pkg/operations/kubernetesupgrade/upgrader.go]: the last line should be `if the topology doesn't have master nodes (so there are no ARM deployments in previous upgradeMasterNodes step) `

Comment (chengliangli0918) [reply pkg/operations/kubernetesupgrade/upgrader.go]: good catch, thank you. Changed in commit 4.

**Pull request merged**